Make transaction id accessible after KTI commit

lutovich opened the pull request:
Transaction id is only assigned during commit of `KernelTransactionImplementation`. It is not assigned if transaction was rolled back or read-only. Currently it is not possible to obtain committed transaction id because KTI just implements `AutoCloseable` and commit happens in `#close()`.

However, it might be sometimes valuable to know committed transaction id and track it.

This commit makes KTI expose it's id (if committed) via `#closeTransaction()` method while still keeping KTI an `AutoCloseable`.

lutovich committed:
Make transaction id accessible after KTI commit

Transaction id is only assigned during commit of KernelTransactionImplementation.
It is not assigned if transaction was rolled back or read-only. Currently it is
not possible to obtain committed transaction id because KTI just implements
AutoCloseable and commit happens in #close().

However, it might be sometimes valuable to know committed transaction id and
track it.

This commit makes KTI expose it's id (if committed) via #closeTransaction()
method while still keeping KTI an AutoCloseable.

lutovich commented:
@davidegrohmann please check this out.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/api/KernelTransaction.java:
```diff
@@ -108,9 +110,20 @@
      * Closes this transaction, committing its changes iff {@link #success()} has been called and
      * {@link #failure()} has NOT been called. Otherwise its changes will be rolled back.
+     *
+     * @return id of the committed transaction or {@link #NOT_COMMITTED} if transaction was rolled back or read-only.
```
I would be nice to have a way to distinguish between rollback and read_only. Maybe we can return -1 for rollback and 0 for read_only?

lutovich replied:
Do you think this will be useful for core-edge or just in general?

davidegrohmann replied:
Yes we do use that information in core-edge RYOW.

burqen replied:
Changes will not be committed if `markForTermination()` has been called, even if `failure()` has not been called explicitly.

davidegrohmann replied:
Termination will eventually perform a rollback, no?

burqen replied:
Sure but that is not what the doc says.
"Closes this transaction, committing its changes if and only if `success()` has been called and `failure()` has NOT been called. Otherwise its changes will be rolled back."
But if `success()` has been called and `failure()` has not, the transaction will still not terminate if `markForTermination()` has been called.

davidegrohmann replied:
Javadoc is lying to you!

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/api/KernelTransaction.java:
```diff
@@ -108,9 +110,20 @@
+    default void close() throws TransactionFailureException
+    {
+        closeTransaction();
+    }
```
I don't really like having two different ways of closing a transaction, but I leave that design decision to you.

lutovich replied:
Yeah, this is a tough for me as well. At the same time it is kind of nice for KTI to be `AutoCloseable`. I guess we can keep it for now and remove `void close()` later if it causes confusion.

davidegrohmann replied:
Fair enough

burqen replied:
Same as for `closeTransaction()`.

burqen commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/integrationtest/KernelIT.java:
```diff
@@ -518,6 +523,76 @@ public void shouldKillTransactionsOnShutdown() throws Throwable
+
+        assertEquals( KernelTransaction.NOT_COMMITTED, tx.closeTransaction() );
+        assertFalse( tx.isOpen() );
+    }
```
Some tests that cover behaviour for terminated transactions would be nice.

lutovich committed:
Differentiate between rollback and read-only for KTI id

This commit makes KernelTransactionImplementation#closeTransaction() return
different values when transaction was rolled back or read only. It also
improves javadoc for KernelTransaction interface.

lutovich commented:
@davidegrohmann @burqen your comments should be addressed now

burqen commented:
@davidegrohmann @lutovich I say 👍

burqen commented:
@davidegrohmann Are you happy with what this gives you for `core-edge`? If so, I'll merge.
